### PR TITLE
Refactor blob management

### PR DESCRIPTION
**Pull request by FedyuninV**

### CODE REVIEW

File: NeoML/include/NeoML/Dnn/Dnn.h [hunk 1]
```diff
@@ -279,6 +275,19 @@ class NEOML_API CBaseLayer : public virtual IObject {
 	// has path "Encoder/LSTM/InputHidden"
 	CString GetPath() const;
 
+	// The following section contains interface for the memory optimization during training
+	// The key idea is that the layer may provide additional information about blobs required for backward and
```
Comment (SAngeliuk) [inline NeoML/include/NeoML/Dnn/Dnn.h]: to be continued...

File: NeoML/src/Dnn/BaseLayer.cpp [hunk 2]
```diff
@@ -428,18 +437,19 @@ void CBaseLayer::runOnce()
 
 		inputBlobs[i] = prevLayerOutput;
 
-		if( GetDnn()->isReuseMemoryMode ) {
+		if( GetDnn()->isReuseMemoryMode && notifyAboutOutput ) {
```
Comment (SAngeliuk) [inline NeoML/src/Dnn/BaseLayer.cpp:440]: Will output blobs be deleted less often?

File: NeoML/src/Dnn/BaseLayer.cpp [hunk 3]
```diff
@@ -756,18 +771,32 @@ void CBaseLayer::CheckOutputs() const
...
 
 	CPtr<CDnnBlob> result = outputBlobs[index];
 	outputProcessedCount[index]++;
-	if( outputProcessedCount[index] == outputs[index] ) {
+	if( outputProcessedCount[index] == outputs[index] && ( blobsNeededForBackward & TOutputBlobs ) == 0 ) {
 		outputBlobs[index] = 0;
 	}
 }
 
+void CBaseLayer::freeUnusedBlobs( int usedBlobs )
```
Comment (SAngeliuk) [inline NeoML/src/Dnn/BaseLayer.cpp]: freeUnused gets used, this is difficult to understand...

File: NeoML/src/Dnn/Layers/CastLayer.cpp [hunk 4]
```diff
@@ -69,7 +69,6 @@ void CCastLayer::RunOnce()
 
 void CCastLayer::BackwardOnce()
 {
-	NeoAssert( inputBlobs[0]->GetDataType() == CT_Float && outputBlobs[0]->GetDataType() == CT_Float );
```
Comment (SAngeliuk) [inline NeoML/src/Dnn/Layers/CastLayer.cpp:72]: May be use diff blobs?

File: NeoML/src/Dnn/BaseLayer.cpp (hunk 2 above)
Comment (FedyuninV) [reply NeoML/src/Dnn/BaseLayer.cpp:440]: Nope.

During inference nothing changes (`notifyAboutOutput` is `true` due to `!GetDnn()->isBackwardPerformed`)

But there is a tricky case during _training_ a pair of layer `FullyConnected->Activation` _inside of recurrent layer_. In this case if we're notifying FC without any restrictions, then the memory consumption may _increase_ (it's difficult to explain here).

File: NeoML/src/Dnn/Layers/CastLayer.cpp (hunk 4 above)
Comment (FedyuninV) [reply NeoML/src/Dnn/Layers/CastLayer.cpp:72]: Fixed by moving check to `::Reshape()` (which is more optimal)

File: NeoML/include/NeoML/Dnn/Dnn.h (hunk 1 above)
Comment (FedyuninV) [reply NeoML/include/NeoML/Dnn/Dnn.h]: fixed

File: NeoML/src/Dnn/BaseLayer.cpp (hunk 3 above)
Comment (FedyuninV) [reply NeoML/src/Dnn/BaseLayer.cpp]: I've made an attempt to fix this... Not sure if it has become any clearer though

**Review by SAngeliuk — APPROVED** [review]

**Pull request merged**